File: tests/test_lesson_analysis.py
import hashlib
import json
import re
import unittest
from pathlib import Path
from xml.etree import ElementTree


ROOT = Path(__file__).resolve().parents[1]
SITE = ROOT / "site"
FIXTURE_PATH = SITE / "data" / "lesson-analysis-svg-mvp.json"
ASSET_ROOT = (
    SITE
    / "assets"
    / "positions"
    / "lesson-analysis-svg-mvp"
    / "opening-fixture"
)
CUBE_LESSON = SITE / "learn" / "cube" / "what-the-cube-is-asking.qmd"
CHECKER_LESSON = (
    SITE
    / "learn"
    / "cube"
    / "why-is-25-percent-the-basic-take-point.qmd"
)


class LessonAnalysisFixtureTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.data = json.loads(FIXTURE_PATH.read_text(encoding="utf-8"))
        cls.cube_source = CUBE_LESSON.read_text(encoding="utf-8")
        cls.checker_source = CHECKER_LESSON.read_text(encoding="utf-8")

    def test_fixture_contract_and_explicit_status(self):
        self.assertEqual(
            self.data["schema_version"],
            "bs-lesson-analysis-fixture-v1",
        )
        self.assertEqual(self.data["fixture_status"]["kind"], "fixture-only")
        message = self.data["fixture_status"]["message"].casefold()
        self.assertIn("test fixtures", message)
        self.assertIn("not verified engine output", message)

    def test_cube_fixture_supports_all_requested_answer_shapes(self):
        cube_cases = self.data["cube_cases"]
        self.assertEqual(
            cube_cases["cube-roll"]["correct_first_action"],
            "roll",
        )
        self.assertEqual(
            cube_cases["cube-double-take"]["actions"]["double"]["responder"][
                "correct_response"
            ],
            "take",
        )
        self.assertEqual(
            cube_cases["cube-double-pass"]["actions"]["double"]["responder"][
                "correct_response"
            ],
            "pass",
        )

    def test_every_referenced_svg_exists_and_parses(self):
        names = set()
        for cube in self.data["cube_cases"].values():
            names.add(cube["initial"]["image"])
            responder = cube["actions"]["double"].get("responder")
            if responder:
                names.add(responder["image"])
        for checker in self.data["checker_cases"].values():
            names.add(checker["initial"]["image"])
            names.update(candidate["image"] for candidate in checker["candidates"])

        self.assertEqual(
            names,
            {
                "starting.svg",
                "responder-flipped.svg",
                "candidate-1.svg",
                "candidate-2.svg",
                "candidate-3.svg",
            },
        )
        for name in names:
            path = ASSET_ROOT / name
            self.assertTrue(path.is_file(), name)
            root = ElementTree.parse(path).getroot()
            self.assertTrue(root.tag.endswith("svg"), name)

    def test_shared_start_is_one_asset_used_by_both_component_types(self):
        cube_start = self.data["cube_cases"]["cube-double-take"]["initial"][
            "image"
        ]
        checker_start = self.data["checker_cases"][
            "checker-three-candidates"
        ]["initial"]["image"]
        self.assertEqual(cube_start, checker_start)
        starts = list(ASSET_ROOT.rglob("starting.svg"))
        self.assertEqual(starts, [ASSET_ROOT / "starting.svg"])

    def test_two_lessons_use_root_relative_fixture_loading(self):
        self.assertIn(
            'data-bs-fixture-src="/data/lesson-analysis-svg-mvp.json"',
            self.cube_source,
        )
        self.assertIn(
            'data-bs-fixture-src="/data/checker-sage-gnu-disagreement-001.json"',
            self.checker_source,
        )
        self.assertEqual(
            self.cube_source.count("data-bs-cube-decision"),
            2,
        )
        self.assertEqual(
            self.checker_source.count("data-bs-checker-decision"),
            1,
        )
        self.assertNotIn("<svg", self.cube_source.casefold())
        self.assertNotIn("<svg", self.checker_source.casefold())

    def test_qmd_hosts_do_not_hard_code_component_ids(self):
        for source in (self.cube_source, self.checker_source):
            host_blocks = re.findall(
                r"<div\s+.*?data-bs-(?:cube|checker)-decision.*?</div>",
                source,
                flags=re.DOTALL,
            )
            self.assertTrue(host_blocks)
            for block in host_blocks:
                self.assertNotRegex(block, r'(?:^|\s)id="')

    def test_script_is_loaded_before_continuous_lesson_loader(self):
        scripts = (SITE / "includes" / "bs-scripts.html").read_text(
            encoding="utf-8"
        )
        analysis_index = scripts.index("bs-lesson-analysis.js")
        scroll_index = scripts.index("bs-learn-scroll.js")
        self.assertLess(analysis_index, scroll_index)
        implementation = (
            SITE / "assets" / "bs-lesson-analysis.js"
        ).read_text(encoding="utf-8")
        self.assertIn("window.BSLearn.mountLesson", implementation)
        self.assertIn("mount(rootElement)", implementation)
        self.assertIn("dataset.bsAnalysisMounted", implementation)

    def test_svg_reuse_cannot_duplicate_inline_ids(self):
        implementation = (
            SITE / "assets" / "bs-lesson-analysis.js"
        ).read_text(encoding="utf-8")
        self.assertIn('element("img", "bs-analysis-position-image")', implementation)
        self.assertIn("img.width = 1200", implementation)
        self.assertIn("img.height = 910", implementation)
        self.assertIn('img.loading = "eager"', implementation)
        self.assertNotIn("fetchSvg", implementation)
        start_hash = hashlib.sha256(
            (ASSET_ROOT / "starting.svg").read_bytes()
        ).hexdigest()
        self.assertEqual(len(start_hash), 64)

    def test_missing_optional_values_are_retained(self):
        candidate = self.data["checker_cases"]["checker-three-candidates"][
            "candidates"
        ][2]
        self.assertIsNone(candidate["winning_probabilities"]["win_gammon"])
        self.assertIsNone(candidate["winning_probabilities"]["lose_gammon"])

    def test_resource_contract_copies_dynamic_assets(self):
        config = (SITE / "_quarto.yml").read_text(encoding="utf-8")
        self.assertIn('"assets/positions/**"', config)
        self.assertIn("data/lesson-analysis-svg-mvp.json", config)
        self.assertIn("data/checker-sage-gnu-disagreement-001.json", config)
        self.assertIn("assets/bs-lesson-analysis.css", config)
        provenance = ASSET_ROOT / "PROVENANCE.txt"
        self.assertTrue(provenance.is_file())
        self.assertFalse((ASSET_ROOT / "PROVENANCE.md").exists())


if __name__ == "__main__":
    unittest.main()

File: tests/test_lesson_analysis.js
const assert = require("node:assert/strict");
const fs = require("node:fs");
const path = require("node:path");

const analysis = require("../site/assets/bs-lesson-analysis.js");

const root = path.resolve(__dirname, "..");
const fixtures = JSON.parse(
  fs.readFileSync(
    path.join(root, "site/data/lesson-analysis-svg-mvp.json"),
    "utf8"
  )
);
const realFixtures = JSON.parse(
  fs.readFileSync(
    path.join(root, "site/data/checker-sage-gnu-disagreement-001.json"),
    "utf8"
  )
);

assert.equal(
  analysis.validateFixtureDocument(fixtures),
  fixtures,
  "the checked-in fixture document is accepted"
);
assert.equal(analysis.validateFixtureDocument(realFixtures), realFixtures);

const doubleTake = fixtures.cube_cases["cube-double-take"];
const rollReview = analysis.cubeDecisionState(doubleTake, "roll");
assert.equal(rollReview.actionAccepted, false);
assert.equal(rollReview.responder, null);
assert.equal(rollReview.actionData.analysis.recommendation, "Double");

const doubleChoice = analysis.cubeDecisionState(doubleTake, "double");
assert.equal(doubleChoice.actionAccepted, true);
assert.equal(doubleChoice.responder.correct_response, "take");
assert.equal(doubleChoice.actionData.analysis.recommendation, "Double");

const passReview = analysis.cubeDecisionState(doubleTake, "double", "pass");
assert.equal(passReview.responseAccepted, false);
assert.equal(passReview.responseData.analysis.recommendation, "Take");

const takeAnswer = analysis.cubeDecisionState(doubleTake, "double", "take");
assert.equal(takeAnswer.responseAccepted, true);
assert.equal(takeAnswer.responseData.analysis.recommendation, "Take");

const rollFixture = fixtures.cube_cases["cube-roll"];
assert.equal(
  analysis.cubeDecisionState(rollFixture, "roll").actionAccepted,
  true,
  "the component supports Roll as the accepted first action"
);
assert.equal(
  analysis.cubeDecisionState(rollFixture, "double").responder,
  null,
  "a rejected Double does not invent a responder decision"
);

const doublePass = fixtures.cube_cases["cube-double-pass"];
assert.equal(
  analysis.cubeDecisionState(doublePass, "double", "pass").responseAccepted,
  true,
  "the component supports Pass as the accepted cube response"
);
assert.equal(
  analysis.cubeDecisionState(doublePass, "double", "take").responseAccepted,
  false
);

assert.throws(
  () => analysis.cubeDecisionState(doubleTake, "beaver"),
  /Double or Roll/
);

const realChecker =
  realFixtures.checker_cases["checker-sage-gnu-disagreement-001"];
const realCandidate = analysis.checkerCandidateState(
  realChecker,
  "candidate-1"
);
assert.equal(realCandidate.label, "8/4");
assert.equal(realCandidate.rank, 1);
assert.equal(realCandidate.equity, -1.615);
assert.equal(realCandidate.equity_loss, 0);
assert.equal(realCandidate.winning_probabilities.lose_gammon, 0.677);
assert.equal(realCandidate.explanation, null);
assert.equal(
  analysis.checkerCandidateIdentityMatches(realChecker, realCandidate),
  true
);
assert.equal(
  analysis.checkerCandidateIdentityMatches(realChecker, {
    ...realCandidate,
    analysis_id: "wrong"
  }),
  false
);
assert.throws(
  () => analysis.cubeDecisionState(doubleTake, "double", "beaver"),
  /Pass or Take/
);

const checker = fixtures.checker_cases["checker-three-candidates"];
const candidate1 = analysis.checkerCandidateState(checker, "candidate-1");
const candidate2 = analysis.checkerCandidateState(checker, "candidate-2");
const candidate3 = analysis.checkerCandidateState(checker, "candidate-3");
assert.equal(candidate1.image, "candidate-1.svg");
assert.equal(candidate2.equity_loss, 0);
assert.equal(candidate3.winning_probabilities.win_gammon, null);
assert.throws(
  () => analysis.checkerCandidateState(checker, "candidate-4"),
  /does not define/
);

assert.equal(analysis.formatEquity(0.093), "+0.093");
assert.equal(analysis.formatEquity(-1), "-1.000");
assert.equal(analysis.formatEquity(null), "Not supplied");
assert.equal(analysis.formatProbability(0.58), "58.0%");
assert.equal(analysis.formatProbability(null), "Not supplied");

assert.equal(
  analysis.assetUrl(fixtures.asset_root, fixtures.cube_cases["cube-roll"].initial.image),
  "/assets/positions/lesson-analysis-svg-mvp/opening-fixture/starting.svg"
);
assert.throws(
  () => analysis.assetUrl(fixtures.asset_root, "../outside.svg"),
  /unsafe/
);

analysis.resetInstanceCounter();
const instanceIds = [
  analysis.nextInstanceId("cube", "cube-double-take"),
  analysis.nextInstanceId("cube", "cube-double-take"),
  analysis.nextInstanceId("checker", "checker-three-candidates")
];
assert.equal(new Set(instanceIds).size, instanceIds.length);
assert.match(instanceIds[0], /^bs-analysis-cube-cube-double-take-1$/);
assert.match(instanceIds[2], /^bs-analysis-checker-checker-three-candidates-3$/);

console.log("lesson analysis fixture logic passed");
